File: hena/tyama_hena15.py
#!/usr/bin/env python
#http://qiita.com/Nabetani/items/cba03c96d1ea55f6e861
#http://nabetani.sakura.ne.jp/hena/ord15elebubo/
import sys
if sys.version_info[0]>=3: raw_input=input

def solve(s):
	v=[int(x,16) for x in s.split('/')]
	a=[ [((v[i]>>(31-j))&1)>0 for j in range(32)]+[False]*2 for i in range(2) ]
	s=''
	i=0
	while i<32:
		if not a[0][i] and not a[1][i]:
			i+=1
		elif a[0][i] and a[1][i]:
			if a[0][i+1]:
				if a[0][i+2]:
					s+='N'
					i+=3
				else:
					s+='R'
					i+=2
			elif a[1][i+1]:
				if a[1][i+2]:
					s+='U'
					i+=3
				else:
					s+='L'
					i+=2
		elif a[0][i]:
			if a[0][i+2]:
				s+='T'
				i+=3
			else:
				s+='Z'
				i+=3
		else:
			if a[0][i+2]:
				s+='S'
				i+=3
			else:
				s+='J'
				i+=2
	print(s)
	sys.stdout.flush()

try:
	while True:
		s=raw_input()
		solve(s)
except EOFError:
	pass

File: hena/tyama_hena15.swift
//usr/bin/env swift $0 $@;exit
//http://qiita.com/Nabetani/items/cba03c96d1ea55f6e861
//http://nabetani.sakura.ne.jp/hena/ord15elebubo/
import CoreFoundation
var v:[UInt]=[0,0] // I cannot specify the array size...
var i=0,j=0
while (withUnsafePointer(to:&v[0]){withVaList([OpaquePointer($0)]){vscanf("%08x",$0)}}>0) {
	_=withUnsafePointer(to:&v[1]){withVaList([OpaquePointer($0)]){vscanf("/%08x",$0)}}
	var a:[[Int]]=[[],[]]
	for i in 0..<2 {
		for j in 0..<32 {
			a[i].append( Int((v[i]>>UInt(31-j))&1) )
		}
		a[i].append(0)
		a[i].append(0)
	}

	i=0
	while i<32 {
		if a[0][i] == 0 && a[1][i] == 0 {
			i+=1
		} else if a[0][i] != 0 && a[1][i] != 0 { //need space; otherwise treated as "a[0][i]!"
			if a[0][i+1] != 0 {
				if a[0][i+2] != 0 {
					print("N",terminator:"")
					i+=3
				} else {
					print("R",terminator:"")
					i+=2
				}
			} else if a[1][i+1] != 0 {
				if a[1][i+2] != 0 {
					print("U",terminator:"")
					i+=3
				}else{
					print("L",terminator:"")
					i+=2
				}
			}
		} else if a[0][i] != 0 {
			if a[0][i+2] != 0 {
				print("T",terminator:"")
				i+=3
			} else { //if a[1][i+2] != 0
				print("Z",terminator:"")
				i+=3
			}
		} else {
			if a[0][i+2] != 0 {
				print("S",terminator:"")
				i+=3
			} else {
				print("J",terminator:"")
				i+=2
			}
		}
	}
	print("")
	fflush(stdout)
}
